### ISSUE TITLE
WebGL fingerprinting deception

### ISSUE THREAD

**Issue by Eloston** [bromite/bromite#131]
**Is your feature request related to privacy?**
Only privacy-related feature requests are considered.

Yes

**Is there a patch available for this feature somewhere?**

Not that I'm aware of.

**Describe the solution you'd like**

Do you know if it's practical to implement a WebGL fingerprinting deception mechanism similar to the one for Canvas image data?

Specifically, for WebGL image hashes as implemented in https://browserleaks.com/webgl

**Describe alternatives you've considered**

N/A

**Comment by csagan5** [comment]
@Eloston I have thought of this in the past, but I quickly dismissed the possibility because of the possible vectorial nature of WebGL renderings. You can have pretty advanced color shifts (and certainly not limited to what we see in that image hash there) and texture usage, so a single pixel change might cause major visual glitches.

However this is just speculation, I have not actually tried and right now by looking at the approach used there perhaps it is possible. If we look at https://browserleaks.com/js/webgl.js towards the end there is:

```javascript

try{
  var j=new Uint8Array(131072);
  if (a.readPixels(0,0,256,128,a.RGBA,a.UNSIGNED_BYTE,j),i=JSON.stringify(j).replace(/,?"[0-9]+":/g,""),""==i.replace(/^{[0]+}$/g,""))
```

And `a` is created as:
```javascript

a=b.getContext("webgl2",{preserveDrawingBuffer:!0})||
b.getContext("experimental-webgl2",{preserveDrawingBuffer:!0})||
b.getContext("webgl",{preserveDrawingBuffer:!0})||
b.getContext("experimental-webgl",{preserveDrawingBuffer:!0})||
b.getContext("moz-webgl",{preserveDrawingBuffer:!0})}
```

So if we get a hold of `readPixels` on all of the above (those available in Chromium), then yes we can just re-use the same subchannel color data randomisation approach on the extracted raster. The only boring/lengthy part would be to implement it for all of the modes supported (not just RGBA) as I did for the `SkImage` case; likely it is using a `SkImage` internally as well, so the code might be reused.

I will give a look to this and report back here, thanks for opening this.

**Comment by Eloston** [comment]
Interesting, thanks for the preliminary research.

Here are my thoughts based on what you found:

> can we intercept the result?

Based on the [documentation on WebGLRenderingContext.readPixels()](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/readPixels), note that the data has to be read into a **new** array; i.e. the data is copied into the JavaScript environment in order to be accessed.

I think we can look into intercepting the JavaScript interface definition for `readPixels`; from there we can determine how we can manipulate the data being retrieved (e.g. during the copying into the JavaScript array, or after it has been copied).

> is Skia really involved?

Skia is purely a 2D graphics library, and this readPixels method seems to be a thin layer around OpenGL calls. I think it's safe to say this isn't the case.

**Comment by csagan5** [comment]
> I think we can look into intercepting the JavaScript interface definition for readPixels; from there we can determine how we can manipulate the data being retrieved (e.g. during the copying into the JavaScript array, or after it has been copied).

Yes we could change the data after it has been copied there, but there must be guarantees that access to the `ArrayBufferView` is exclusively locked at that time.

> Skia is purely a 2D graphics library, and this readPixels method seems to be a thin layer around OpenGL calls. I think it's safe to say this isn't the case.

See https://github.com/chromium/chromium/blob/2ca8c5037021c9d2ecc00b787d58a31ed8fc8bcb/gpu/skia_bindings/gl_bindings_skia_cmd_buffer.h

**Comment by Eloston** [comment]
> See https://github.com/chromium/chromium/blob/2ca8c5037021c9d2ecc00b787d58a31ed8fc8bcb/gpu/skia_bindings/gl_bindings_skia_cmd_buffer.h

It would be a bit strange to me if they're converting to Skia first, since `readPixels` seems like a thin layer around the GL calls as I mentioned earlier.

Nevertheless, I would need to start looking at code too in order to contribute something meaningful to this discussion. I can't guarantee I will have the chance to do so anytime soon, though.

**Comment by csagan5** [comment]
> Based on the [documentation on WebGLRenderingContext.readPixels()](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/readPixels), note that the data has to be read into a **new** array; i.e. the data is copied into the JavaScript environment in order to be accessed.
> 
> I think we can look into intercepting the JavaScript interface definition for `readPixels`;

@Eloston the corresponding IDL functions should be here: https://github.com/chromium/chromium/blob/bf31bedf12761be9d4e56500559f2777cbcbeedd/third_party/blink/renderer/modules/webgl/webgl2_rendering_context_base.cc#L734

Interception should be possible right after the call to `ReadPixelsHelper`, or better inside that function because we do not want to edit pixels if no pixel data was added; specifically here: https://github.com/chromium/chromium/blob/53f3d30ef104e4c91ab0d7eb9683305139072ef3/third_party/blink/renderer/modules/webgl/webgl_rendering_context_base.cc#L4345

If you check that file you will find several calls like `ContextGL()->ReadPixels()`, that `ReadPixels` with uppercase `R` comes from the context, like in the case of GLES2 here:
https://github.com/chromium/chromium/blob/2ca8c5037021c9d2ecc00b787d58a31ed8fc8bcb/gpu/command_buffer/client/gles2_implementation.cc#L3808

Next step would be to check which pixels format are maximally supported, so that we know which ones to implement.

**Comment by DecodeMehh** [comment]
Also check the font fingerprint
https://browserleaks.com/fonts

**Comment by csagan5** [comment]
Related: https://bugzilla.mozilla.org/show_bug.cgi?id=1428034

**Comment by baybal** [comment]
I would consider adding an option for turning off WebGL entirely. As much for it having no real use aside for obscene 3D popups, as for near impossibility of sanitising the shader code.

Somewhere in 7x.x development, Google disabled the flag to disable the WebGL, so now I believe we have to put it back.

**Comment by PoorPocketsMcNewHold** [comment]
> I would consider adding an option for turning off WebGL entirely. As much for it having no real use aside for obscene 3D popups, as for near impossibility of sanitising the shader code.
> 
> Somewhere in 7x.x development, Google disabled the flag to disable the WebGL, so now I believe we have to put it back.

However, this could provoke a raise in the uniqueness of your fingerprint as a "not supported/broken webgl" support.

**Comment by csagan5** [comment]
Flags for turning off WebGL have been added, see #411; they are not disabled by default in Bromite.

> this could provoke a raise in the uniqueness of your fingerprint as a "not supported/broken webgl" support.

@PoorPocketsMcNewHold by disabling WebGL/WebGL2 you are swapping *n* bits (whichever amount of entropy they convey) with 1 bit, effectively reducing the leak. This bit will however be used in combination with all the other sources, so that in the best case users can be identified as Bromite users and in the worst case as detailed as any Chromium browser allows (even individually).

**Comment by DirtyAngelica** [comment]
Some things I noticed:
- Disabling WebGL in flags doesn't seem to do anything for detection and/or fingerprinting. BrowserLeaks still detects the hardware renderer and the GPU.
- WebRTC device ID's are not hidden.

Some questions:
- Any chance for Bromite for Windows? It would be a heavy competitor for Brave.
- Why disable QUIC by default? Does it pose a security and/or fingerprint risk?
- On non-rooted Nougat devices, default Android WebView is disabled and Chrome is set as the default Web View. If Chrome is disabled, Nougat auto-selects "Google Web View" (whatever that is). If Chrome is disabled, Bromite WebView is installed, then does Google WebView use Bromite WebView? Nougat is special when it comes to WebView thing.

**Comment by DirtyAngelica** [comment]
Thanks!
FYI, on original rooted Pixel + LineageOS 17.1 Nightly (Android 10) + Magisk + Magisk Bromite Module, Bromite WebView works perfectly, does not break anything, can be selected in Developer Options. Android WebView can be removed fully. Its awesome!

**Comment by sergiok707** [comment]
> Flags for turning off WebGL have been added, see #411; they are not disabled by default in Bromite.
> 
> > this could provoke a raise in the uniqueness of your fingerprint as a "not supported/broken webgl" support.
> 
> @PoorPocketsMcNewHold by disabling WebGL/WebGL2 you are swapping _n_ bits (whichever amount of entropy they convey) with 1 bit, effectively reducing the leak. This bit will however be used in combination with all the other sources, so that in the best case users can be identified as Bromite users and in the worst case as detailed as any Chromium browser allows (even individually).

Hello, i don't find WebGL OFF flag in Bromite 85.0.4183.86
Can you please indicate the exact path of how to find it?

**Comment by uazo** [comment]
I think I've found a way, or at least I'm getting closer...
basically I work the resulting image array always as `GL_RGBA` `GL_UNSIGNED_BYTE` (so I know how it is done in memory) and only after I transform it into the type required by javascript.
since I speak English badly I tell you with the code:

```
diff --git a/third_party/blink/renderer/modules/webgl/webgl_rendering_context_base.cc b/third_party/blink/renderer/modules/webgl/webgl_rendering_context_base.cc
index bba2cc67f1..01bd0b453a 100644
--- a/third_party/blink/renderer/modules/webgl/webgl_rendering_context_base.cc
+++ b/third_party/blink/renderer/modules/webgl/webgl_rendering_context_base.cc
@@ -132,6 +132,9 @@
 #include "third_party/blink/renderer/platform/wtf/threading_primitives.h"
 #include "ui/gfx/geometry/size.h"
 
+#include "third_party/blink/renderer/platform/graphics/gpu/webgl_image_conversion.h"
+#include "base/rand_util.h"
+
 namespace blink {
 
 bool WebGLRenderingContextBase::webgl_context_limits_initialized_ = false;
@@ -4679,7 +4682,9 @@ void WebGLRenderingContextBase::readPixels(
   ReadPixelsHelper(x, y, width, height, format, type, pixels.Get(), 0);
 }
 
-void WebGLRenderingContextBase::ReadPixelsHelper(GLint x,
+#define shuffleComponent(color, max, delta) ( (color) >= (max) ? ((max)-(delta)) : ((color)+(delta)) )
+
+void WebGLRenderingContextBase::ReadPixelsHelper(GLint x,
                                                  GLint y,
                                                  GLsizei width,
                                                  GLsizei height,
@@ -4750,7 +4755,54 @@ void WebGLRenderingContextBase::ReadPixelsHelper(GLint x,
 
   {
     ScopedDrawingBufferBinder binder(GetDrawingBuffer(), framebuffer);
-    ContextGL()->ReadPixels(x, y, width, height, format, type, data);
+
+    uint32_t pixels = (width-x) * (height-y);
+    Vector<uint8_t> data_rgba;
+    data_rgba.resize(pixels*4);
+    uint8_t* data_ptr = data_rgba.data();
+
+    ContextGL()->ReadPixels(x, y, width, height, GL_RGBA, GL_UNSIGNED_BYTE,
+                            data_ptr);
+
+    // calculate randomisation amounts for each RGB component
+    double shuffleX = base::RandDouble();
+    double shuffleY = base::RandDouble();
+
+    // calculate random coordinates using bisection
+    auto currentW = width, currentH = height;
+
+    uint8_t* source = data_ptr;
+    for (unsigned i = 0; i < pixels; ++i) {
+      int x = currentW * shuffleX, y = currentH * shuffleY;
+
+      // calculate randomisation amounts for each RGB component
+      uint8_t shuffleR = base::RandInt(0, 4);
+      uint8_t shuffleG = (shuffleR + x) % 4;
+      uint8_t shuffleB = (shuffleG + y) % 4;
+
+      // apply
+      source[0] = shuffleComponent(source[0], UINT8_MAX-1, shuffleR);
+      source[1] = shuffleComponent(source[1], UINT8_MAX-1, shuffleG);
+      source[2] = shuffleComponent(source[2], UINT8_MAX-1, shuffleB);
+      source += 4;
+
+      // keep bisecting or reset current width/height as needed
+      if (x == 0) {
+        currentW = width;
+      } else {
+        currentW = x;
+      }
+      if (y == 0) {
+        currentH = height;
+      } else {
+        currentH = y;
+      }
+    }
+
+    WebGLImageConversion::DataFormat destinationFormat =
+              WebGLImageConversion::GetWebGLDataFormat(format,type);
+    WebGLImageConversion::PackPixels(data_ptr, destinationFormat,
+              pixels, data);
   }
 }
 
diff --git a/third_party/blink/renderer/platform/graphics/gpu/webgl_image_conversion.cc b/third_party/blink/renderer/platform/graphics/gpu/webgl_image_conversion.cc
index 76a005a168..6bbac30d87 100644
--- a/third_party/blink/renderer/platform/graphics/gpu/webgl_image_conversion.cc
+++ b/third_party/blink/renderer/platform/graphics/gpu/webgl_image_conversion.cc
@@ -4099,6 +4099,12 @@ bool WebGLImageConversion::PackPixels(
   return true;
 }
 
+WebGLImageConversion::DataFormat
+  WebGLImageConversion::GetWebGLDataFormat(GLenum destination_format,
+                                      GLenum destination_type) {
+  return GetDataFormat(destination_format, destination_type);
+}
+
 void WebGLImageConversion::UnpackPixels(const uint16_t* source_data,
                                         DataFormat source_data_format,
                                         unsigned pixels_per_row,
diff --git a/third_party/blink/renderer/platform/graphics/gpu/webgl_image_conversion.h b/third_party/blink/renderer/platform/graphics/gpu/webgl_image_conversion.h
index ef574ea3b5..a7c049206e 100644
--- a/third_party/blink/renderer/platform/graphics/gpu/webgl_image_conversion.h
+++ b/third_party/blink/renderer/platform/graphics/gpu/webgl_image_conversion.h
@@ -263,6 +263,14 @@ class PLATFORM_EXPORT WebGLImageConversion final {
 
   // End GraphicsContext3DImagePacking.cpp functions
 
+  static void PackPixels(const uint8_t* source_data,
+                         DataFormat source_data_format,
+                         unsigned pixels_per_row,
+                         uint8_t* destination_data);
+
+  static DataFormat GetWebGLDataFormat(GLenum destination_format,
+                                       GLenum destination_type);
+
  private:
   friend class WebGLImageConversionTest;
   // Helper for packImageData/extractImageData/extractTextureData, which
@@ -288,10 +296,6 @@ class PLATFORM_EXPORT WebGLImageConversion final {
                            DataFormat source_data_format,
                            unsigned pixels_per_row,
                            uint8_t* destination_data);
-  static void PackPixels(const uint8_t* source_data,
-                         DataFormat source_data_format,
-                         unsigned pixels_per_row,
-                         uint8_t* destination_data);
 };
 
 }  // namespace blink
```
tell me if it can work and how can I improve it.
could you tell me some url for testing, to date I have used https://www.bromite.org/detect https://browserleaks.com/webgl https://abrahamjuliot.github.io/creepjs/ but I would need something that from the readpixels would show me the resulting image

**Comment by csagan5** [comment]
> tell me if it can work and how can I improve it.

I see you have added this [here](https://github.com/uazo/bromite/blob/next/v99/build/patches/Add-site-settings-for-webgl.patch).

Only problem I see is that this leads to:
1. re-encode into `GL_RGBA` `GL_UNSIGNED_BYTE`
2. re-encode into original format

So it could be a quite sensible performance hit; perhaps we can create one function for each format? I doubt it can be done with C++ templating.

**Comment by uazo** [comment]
> I see you have added this [here](https://github.com/uazo/bromite/blob/next/v99/build/patches/Add-site-settings-for-webgl.patch).

yes, attention that [this](https://github.com/uazo/bromite/blob/next/v99/build/patches/Content-settings-infrastructure.patch) is also needed 

> So it could be a quite sensible performance hit; 

no, the transformation into `GL_RGBA` + `GL_UNSIGNED_BYTE` it's a simple array copy

> I doubt it can be done with C++ templating.

actually yes, you can find them all [here](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/graphics/gpu/webgl_image_conversion.cc;)

**Comment by csagan5** [comment]
> > I see you have added this [here](https://github.com/uazo/bromite/blob/next/v99/build/patches/Add-site-settings-for-webgl.patch).
> 
> yes, attention that [this](https://github.com/uazo/bromite/blob/next/v99/build/patches/Content-settings-infrastructure.patch) is also needed

I am not adding any of these patches until further discussion happens.

> > So it could be a quite sensible performance hit;
> 
> no, the transformation into `GL_RGBA` + `GL_UNSIGNED_BYTE` it's a simple array copy

You are assuming that this has no performance impact but it will have one for webpages that render many frames per second.

> > I doubt it can be done with C++ templating.
> 
> actually yes, you can find them all [here](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/graphics/gpu/webgl_image_conversion.cc;)

There are no such functions there. I am talking about "we can create one function for each format" -> one function that adds noise for each pixel format.

**Comment by uazo** [comment]
> but it will have one for webpages that render many frames per second.

which?

> I am talking about "we can create one function for each format"

precisely. there are all the formats available and how they are organized in memory, and the code is a C++ template

**Comment by csagan5** [comment]
> > but it will have one for webpages that render many frames per second.
> 
> which?

Try any website which does high frequency updates, for example a game.

**Comment by uazo** [comment]
> Try any website which does high frequency updates, for example a game.

I tried all of the ones on https://webglsamples.org/ and they worked well

**Comment by csagan5** [comment]
If you find a device which is not so powerful, it will have an impact.

For the canvas we have a function per pixel type so we should do the same here.

**Comment by uazo** [comment]
> If you find a device which is not so powerful, it will have an impact.

but are we talking about the same thing? I'm talking about the `ReadPixel` function...
we are interested in the uses not proper to the API, if a game, for example, works badly, being a content setting the user can simply modify it. but in the tests I did, I did not find any lag.

also, a little [research](https://sourcegraph.com/search?q=context:global+ReadPixels%28+AND+lang:javascript+&patternType=regexp) shows that sites use `RGBA UNSIGNED_BYTE` which, being the default, does not perform any data transformation. developing it for every type of image may not be worth it.
it is true, there are a few examples, but unfortunately I don't have access to big query maybe from there we would see more uses of that api.

**Comment by csagan5** [comment]
Yes: see https://github.com/bromite/bromite/blob/master/build/patches/Multiple-fingerprinting-mitigations.patch#L513

There is a case switch for every pixel format type. The WebGL version should do the same to avoid the double conversion.

**Comment by uazo** [comment]
> There is a case switch for every pixel format type

yes I had seen that function, but from what I have seen not all modes are managed and maybe it should be revised:

[ImageData::GetSkPixmap()](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/canvas/image_data.cc;l=349;drc=e2cba64c183ae17816143ee344e6f7c81451555a;bpv=1;bpt=1)
[enum SkColorType](https://source.chromium.org/chromium/chromium/src/+/main:third_party/skia/include/core/SkImageInfo.h;drc=e2cba64c183ae17816143ee344e6f7c81451555a;bpv=1;bpt=1;l=64)

> The WebGL version should do the same to avoid the double conversion.

technically there is no double conversion, the pack functions always start from `RGBA UNSIGNED_BYTE` which makes me think it's the de-facto default. unfortunately it is complex to prove, since the calls to the graphics driver are difficult to follow.
eventually we could verify the impact on performance with some tests, but honestly I think the best way is the one I propose, it assures us that any new formats are handled by chromium with minimal effort

**Comment by csagan5** [comment]
> yes I had seen that function, but from what I have seen not all modes are managed and maybe it should be revised:
> 
> [ImageData::GetSkPixmap()](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/canvas/image_data.cc;l=349;drc=e2cba64c183ae17816143ee344e6f7c81451555a;bpv=1;bpt=1) [enum SkColorType](https://source.chromium.org/chromium/chromium/src/+/main:third_party/skia/include/core/SkImageInfo.h;drc=e2cba64c183ae17816143ee344e6f7c81451555a;bpv=1;bpt=1;l=64)

Time to add more pixel modes then.

> technically there is no double conversion, the pack functions always start from `RGBA UNSIGNED_BYTE` which makes me think it's the de-facto default. unfortunately it is complex to prove,

You can add a log line and then from Javascript use a different pixel format.

> honestly I think the best way is the one I propose, it assures us that any new formats are handled by chromium with minimal effort

I disagree, we cannot have a different approach for the Canvas and WebGL.

**Comment by csagan5** [comment]
> > Time to add more pixel modes then.
> 
> reading the modalities seems extremely complex to me, in the meantime we are exposed, correct?

There is nothing exposed, these are just mitigations. Mitigations would not cover all pixel formats, but that's it.

> > You can add a log line and then from Javascript use a different pixel format.
> 
> I do not understand

Add a log line to report which pixel format is being used in Chromium code, then from Javascript request a WebGL buffer in different formats: https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/readPixels
And verify whether the internal Chromium code only uses one format (RGBA) or not.

**Comment by JustAnUserLol** [comment]
WebGL has some security risks too; Can you try to make WebGL not only private But secure also?

**Comment by uazo** [comment]
> Add a log line to report which pixel format is being used in Chromium code, then from Javascript request a WebGL buffer in different formats

I have checked and I have come to the conclusion that it makes no sense to proceed.
in whatever way I do, what I produce is an incomplete webgl, which works yes, but not always and not for everything. so what's the use of enabling it halfway?
and it is not a performance problem (readPixel is not used when rendering), but rather since the information that the library can give to javascript is too much, and therefore denying some information and obfuscating others always gets to the point that I'm denying functionality.
all the better to deactivate it and if the user wants to activate it, proceed at his own risk.

So, do you agree if I introduce a simple on/off content setting for webgl?

**Comment by csagan5** [comment]
> in whatever way I do, what I produce is an incomplete webgl, which works yes, but not always and not for everything. so what's the use of enabling it halfway?

You mentioned that it always starts from `RGBA UNSIGNED_BYTE`; I still do not know if that is the case or not. Even if it where, we could implement some pixel formats and have a non-optimized fallback.
Why would anything be incomplete/not working? The purpose of this change is not to break WebGL but to obfuscate the color data.

> and it is not a performance problem (readPixel is not used when rendering), but rather since the information that the library can give to javascript is too much, and therefore denying some information and obfuscating others always gets to the point that I'm denying functionality.

I do not know what you are referring to but here we were discussing only obfuscating the pixel color data, not removing any functionality.

> So, do you agree if I introduce a simple on/off content setting for webgl?

There is already a flag for this but sure, we could have it as a content setting so that at least WebGL will not be used opportunistically by sites that should have no use for it.

**Comment by uazo** [comment]
> You mentioned that it always starts from RGBA UNSIGNED_BYTE;  I still do not know if that is the case or not

no, this is not the case, because it depends on the driver.
in webgl there is the possibility to understand which formats can be used through `IMPLEMENTATION_COLOR_READ_FORMAT` and `IMPLEMENTATION_COLOR_READ_TYPE`, but no way to understand what the default is, although I am convinced by reading the code that that is the standard, but I have no proof

> The purpose of this change is not to break WebGL but to obfuscate the color data.

I started from the idea that the goal was to minimize the possibility of use for fingerprinting: but changing the color is just one of them.
I found so much information returned to javascript that I think that changing the color alone would give a false impression of privacy, so I say it's not worth it.
you should check every single info and understand if it can be used for fingerprinting, but that code is hard, the standard in evolution and above all it is closed when I get to talk to the driver.
I thought that the code of brave was complete but unfortunately they too have no longer aligned it with the subsequent modifications of the webgl standard, so it is unusable as is.

> There is already a flag for this but sure, we could have it as a content setting so that at least WebGL will not be used opportunistically by sites that should have no use for it.

exactly my idea.
I think I will also check the code of the canvas and the blobs, my fear is that even there it would be better to deactivate reading data from javascript than to simple change the color (especially when I read in google js code this error message "Post is not supported when using canvas url" and "Wrong origin for canvas or hidden url" :)

**Comment by csagan5** [comment]
> > You mentioned that it always starts from RGBA UNSIGNED_BYTE;  I still do not know if that is the case or not
> 
> no, this is not the case, because it depends on the driver. in webgl there is the possibility to understand which formats can be used through `IMPLEMENTATION_COLOR_READ_FORMAT` and `IMPLEMENTATION_COLOR_READ_TYPE`, but no way to understand what the default is, although I am convinced by reading the code that that is the standard, but I have no proof

The idea of logging was to understand how it works.



> > The purpose of this change is not to break WebGL but to obfuscate the color data.
> 
> I started from the idea that the goal was to minimize the possibility of use for fingerprinting: but changing the color is just one of them.

This issue is about the WebGL canvas because a few online sites (and libraries) use the WebGL canvas for fingerprinting; this issue is not a catch-all for any type of fingerprinting.

> I found so much information returned to javascript (and different ways to get them) that I think that changing the color alone would give a false impression of privacy, so I say it's not worth it.

It is worth it because it is one of the most common methods used for fingerprinting. Any counter-measure should be based on the study of existing fingerprinting libraries (both open source and commercial), not on sites that show "samples" of what is possible. In those cases I agree that they give fake reassurance to users.

> you should check every single info and understand if it can be used for fingerprinting, but that code is hard, the standard in evolution and above all it is closed when I get to talk to the driver.

The best approach is to consider giving WebGL permissions only to some trusted sites; it is a security and privacy nightmare otherwise.

> I think I will also check the code of the canvas and the blobs, my fear is that even there it would be better to deactivate reading data from javascript than to simple change the color (especially when I read in google js code this error message "Post is not supported when using canvas url" and "Wrong origin for canvas or hidden url" :)

What are these about?

**Comment by uazo** [comment]
> What are these about?

I am trying to read google javascript code to exploit it with the aim of creating a userscript (and possibly enhance the bromite support of it, such as hiding method overrides by pretending they are native, as a mechanism for deceiving fingerprinting methods), and I found those strings. if I get some more information I will open a discussion.

**Comment by Stoptimerag** [comment]
Some website use readPixal(),but some website use toDataUrl() or toBlob().
more information:[https://fingerprints.bablosoft.com/webgl](url)
